File: Hash Table/454. 4Sum II.cpp
/*
454. 4Sum II

Given four lists A, B, C, D of integer values, compute how many tuples (i, j, k, l) 
there are such that A[i] + B[j] + C[k] + D[l] is zero.

To make problem a bit easier, all A, B, C, D have same length of N where 0 ≤ N ≤ 500.
All integers are in the range of -228 to 228 - 1 and the result is guaranteed to be at most 231 - 1.

Example:

Input:
A = [ 1, 2]
B = [-2,-1]
C = [-1, 2]
D = [ 0, 2]

Output:
2

Explanation:
The two tuples are:
1. (0, 0, 0, 1) -> A[0] + B[0] + C[0] + D[1] = 1 + (-2) + (-1) + 2 = 0
2. (1, 1, 0, 0) -> A[1] + B[1] + C[0] + D[0] = 2 + (-1) + (-1) + 0 = 0

*/


class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        unordered_map<int,int>AB_sum;
        for(auto a: A)
            for(auto b: B)
                AB_sum[a+b]++;
        int result = 0; 
        for(auto c: C)
            for(auto d: D)
                if(AB_sum.find(-c-d)!=AB_sum.end())
                    result += AB_sum[-c-d];
        return result;
    }
};

File: Hash Table/454. 4Sum II.py
"""
454. 4Sum II

Input:
A = [ 1, 2]
B = [-2,-1]
C = [-1, 2]
D = [ 0, 2]

Output:
2

Explanation:
The two tuples are:
1. (0, 0, 0, 1) -> A[0] + B[0] + C[0] + D[1] = 1 + (-2) + (-1) + 2 = 0
2. (1, 1, 0, 0) -> A[1] + B[1] + C[0] + D[0] = 2 + (-1) + (-1) + 0 = 0

"""

class Solution:
    def fourSumCount(self, A, B, C, D):
        dic = collections.Counter([a+b for a in A for b in B])
        return sum([dic[-c-d] for c in C for d in D if -c-d in dic])

class Solution:
    def fourSumCount(self, A, B, C, D):
        """
        :type A: List[int]
        :type B: List[int]
        :type C: List[int]
        :type D: List[int]
        :rtype: int
        """
        dic = collections.defaultdict(int)
        for a in A:
            for b in B:
                dic[a+b]+=1
        res = 0;
        for c in C: 
            for d in D:
                if -c-d in dic:
                    res += dic[-c-d]
        return res